proposals/007-azure-kms.md

```diff
@@ -0,0 +1,148 @@
+
+Public addresses like the Key Vault base url will be different per national cloud. For example in China an address might look like `https://vault-name.vault.azure.cn`. This also impacts the `scope` we request via the oauth client credentials flow, we would likely want to request `https://vault.azure.cn/.default` scope.
+
+Entra oauth endpoint will vary by National Cloud
+
+```
+Global   https://login.microsoftonline.com
+Microsoft Entra ID for US Government	https://login.microsoftonline.us
+Microsoft Entra China operated by 21Vianet	https://login.partner.microsoftonline.cn
+...
+```
+
+The implication for us is that the oauth endpoint and scope should be configurable. The vault base uri is a natural part of the configuration anyway. We could potentially infer the `scope` from the configured vault base uri.
+
+[National clouds](https://learn.microsoft.com/en-us/entra/identity-platform/authentication-national-cloud)
+
+### TLS
+
+With two remote systems involved we want independently customizable TLS for:
+
+1. the authentication client
+2. the key vault client
+
+So that we can supply custom trust or use insecure testing modes (mock servers using self-signed certs for example).
+
+### DEK Generation
+
+Another difference is that basic Key Vault cannot generate the DEK bytes for us. Managed HSM does have an operation to generate random bytes for us https://learn.microsoft.com/en-us/rest/api/keyvault/?view=rest-keyvault-keys-7.4#key-operations-managed-hsm-only
+
+### Key Identifiers
+
+When we wrap/unwrap we have to identify a key by its name and version. You can obtain the latest version by https://learn.microsoft.com/en-us/rest/api/keyvault/keys/get-key/get-key?view=rest-keyvault-keys-7.4&tabs=HTTP.
+
+The API usually returns it as a `kid` https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#object-identifiers. Like
+
+> For Vaults: https://{vault-name}.vault.azure.net/{object-type}/{object-name}/{object-version}
+
+eg. `https://my-vault.vault.azure.net/keys/my-key/78deebed173b48e48f55abf87ed4cf71`
+
+The Azure SDK then takes advantage of the fact this happens to be the base URL for various operations like wrap/unwrap.
+
+ed. for [wrap key](https://learn.microsoft.com/en-us/rest/api/keyvault/keys/wrap-key/wrap-key?view=rest-keyvault-keys-7.4&tabs=HTTP) the url is `POST {vaultBaseUrl}/keys/{key-name}/{key-version}/wrapkey?api-version=7.4
+`
+
+To minimise EDEK bytes there are a couple of opportunities when encoding the EDEK:
+1. we can exclude the vaultBaseUrl and have the limitation that we work only with a single key vault
+2. we can encode just the keyName and keyVersion, and use that to decrypt
+3. all documented examples of the keyVersion (and from my experimentation too) are 128bits encoded as a hexadecimal string.
```

Comment:
I'd be wary of relying on the results of experimentation.  If Azure change the naming scheme we risk a user blowing up.    Unless we can get a statement from Microsoft saying it is 128bits and hex, I think we need to be safe are treat it as a string.  Annoying.  

We had the same situation in AWS, the keyid looks like a UUID but they are documented as Strings.  We treat them as Strings. Grrr.

Reply:
what I've got in the prototype is I try to optimistically treat it as a 16-byte hex string and fall back to arbitrary-string, so:
* if we can decode the id as hex and it produces 16 bytes, we serialize a compact EDEK with version byte 0
* else serialize a stringy EDEK with version byte 1

I also think we can use a single byte for the string length as it's docced as being 32 characters, I don't think it can blow out to more than 127 bytes even if it contained all obscure unicode characters.

Reply:
or you could come up with some scheme to do the same thing within a single message version. like have a 'versionType' field that tells you how to interpret the bytes.

but yeah we have to be very careful, I'd only want to apply this if we can guarantee fidelity. Like Azure says the object ids are case insensitive, but I'd be wary of encoding a hex string containing uppercase alpha characters as we lose the case information.